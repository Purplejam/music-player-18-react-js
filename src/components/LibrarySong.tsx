 import {SongType} from '../App';
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
 import { faCirclePlay, faVolumeUp, faHeartCircleCheck } from '@fortawesome/free-solid-svg-icons';


//types
 type LibrarySongType = {
	 songs: SongType[], 
		setSong: Function, 
		song: SongType, 
		setCurrentSong: Function, 
		setPlaying: Function, 
		isPlaying: boolean, 
		audioRef: any,
		currentSong: SongType
 }

//icons
const playIcon = <FontAwesomeIcon icon={faCirclePlay} />
const volumeIcon = <FontAwesomeIcon icon={faVolumeUp} />
const likeIcon = <FontAwesomeIcon data-testid="like-icon" icon={faHeartCircleCheck} />

function LibrarySong({
	songs, 
	setSong, 
	currentSong,
	song, 
	setCurrentSong, 
	setPlaying, 
	isPlaying, 
	audioRef}: LibrarySongType) {

//set active song with click
function setActiveSong(): void {
	const newSongs = songs.map((item) => {
		if (item.id === song.id) {
			return {
				...item,
				active: true
			}
		} else {
						return {
				...item,
				active: false
			}
		}
	})
	setSong(newSongs);
}

//event handlers
	async function songSelectHandler(): Promise<void> {
		if (isPlaying) {
			setPlaying(false);
			await setCurrentSong(song);
			setPlaying(true);
			audioRef.current.play();
			setActiveSong();
		} 
		if (!isPlaying) {
			await setCurrentSong(song);
			setActiveSong();
		}
	}

	return(
		<div className={`${song.active ? "selected" : ''} library-song`} onClick={songSelectHandler}>
			<img src={song.cover} alt={song.name}/>
			<div className="song-description">
				<h3 className={`${song.favorite ? "favorite" : ''}`}>{song.name}<i>{likeIcon}</i></h3>
				<h4>{song.artist}</h4>
			</div>
			<span className="library-song-icon">{currentSong.id === song.id && isPlaying ? volumeIcon : playIcon}</span>
		</div>
		)
}


export default LibrarySong;