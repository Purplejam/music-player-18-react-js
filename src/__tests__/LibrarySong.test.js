import {render, screen} from '@testing-library/react';
import LibrarySong from '../components/LibrarySong';
import { v4 as uuidv4 } from "uuid";

const mockSong = {
    name: 'The Greatest Song',
    cover: 'Fantastic Image',
    artist: 'Popular Artist',
    audio: 'link',
    color: ['#ffffff', '#000000'],
    id: uuidv4(),
    active: false,
    favorite: true,
}

describe('<LibrarySong/>', () => {
	it('renders properly a custom song', () => {
		const {container, debug} = render(
			<LibrarySong currentSong={mockSong} song={mockSong}/>
			)
		const nameText = screen.getByText('The Greatest Song');
		expect(nameText).toBeInTheDocument();

		const artistText = screen.getByText('Popular Artist');
		expect(artistText).toBeInTheDocument();
	})

	it('renders proper icon if it`s favorite song', () => {
		const {container, debug} = render(
			<LibrarySong currentSong={mockSong} song={mockSong}/>
			)

		const icon = screen.getByTestId('like-icon');
		expect(icon).toHaveAttribute('data-icon', 'heart-circle-check');
	})

	it('matches the snapshot', () => {
		const {container, debug} = render(
			<LibrarySong currentSong={mockSong} song={mockSong}/>
			)	
			expect(container).toMatchSnapshot();
	})
})

