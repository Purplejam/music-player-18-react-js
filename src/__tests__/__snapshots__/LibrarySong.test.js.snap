// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LibrarySong/> matches the snapshot 1`] = `
<div>
  <div
    class=" library-song"
  >
    <img
      alt="The Greatest Song"
      src="Fantastic Image"
    />
    <div
      class="song-description"
    >
      <h3
        class="favorite"
      >
        The Greatest Song
        <i>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart-circle-check "
            data-icon="heart-circle-check"
            data-prefix="fas"
            data-testid="like-icon"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 368C256 403.7 266.6 436.9 284.9 464.6L279.4 470.3C266.4 483.2 245.5 483.2 233.5 470.3L39.71 270.5C-16.22 212.5-13.23 116.6 49.7 62.68C103.6 15.73 186.5 24.72 236.5 75.67L256.4 96.64L275.4 75.67C325.4 24.72 407.3 15.73 463.2 62.68C506.1 100.1 520.7 157.6 507 208.7C484.3 198 458.8 192 432 192C334.8 192 256 270.8 256 368zM576 368C576 447.5 511.5 512 432 512C352.5 512 288 447.5 288 368C288 288.5 352.5 224 432 224C511.5 224 576 288.5 576 368zM476.7 324.7L416 385.4L387.3 356.7C381.1 350.4 370.9 350.4 364.7 356.7C358.4 362.9 358.4 373.1 364.7 379.3L404.7 419.3C410.9 425.6 421.1 425.6 427.3 419.3L499.3 347.3C505.6 341.1 505.6 330.9 499.3 324.7C493.1 318.4 482.9 318.4 476.7 324.7H476.7z"
              fill="currentColor"
            />
          </svg>
        </i>
      </h3>
      <h4>
        Popular Artist
      </h4>
    </div>
    <span
      class="library-song-icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-play "
        data-icon="circle-play"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256C0 114.6 114.6 0 256 0C397.4 0 512 114.6 512 256zM176 168V344C176 352.7 180.7 360.7 188.3 364.9C195.8 369.2 205.1 369 212.5 364.5L356.5 276.5C363.6 272.1 368 264.4 368 256C368 247.6 363.6 239.9 356.5 235.5L212.5 147.5C205.1 142.1 195.8 142.8 188.3 147.1C180.7 151.3 176 159.3 176 168V168z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Nav/> matches the snapshot 1`] = `
<div>
  <nav>
    <a
      data-testid="logo-icon"
      href="/"
    >
      <h1>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-compact-disc "
          data-icon="compact-disc"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 0C114.6 0 0 114.6 0 256s114.6 256 256 256s256-114.6 256-256S397.4 0 256 0zM80.72 256H79.63c-9.078 0-16.4-8.011-15.56-17.34C72.36 146 146.5 72.06 239.3 64.06C248.3 63.28 256 70.75 256 80.09c0 8.35-6.215 15.28-14.27 15.99C164.7 102.9 103.1 164.3 96.15 241.4C95.4 249.6 88.77 256 80.72 256zM256 351.1c-53.02 0-96-43-96-95.1s42.98-96 96-96s96 43 96 96S309 351.1 256 351.1zM256 224C238.3 224 224 238.2 224 256s14.3 32 32 32c17.7 0 32-14.25 32-32S273.7 224 256 224z"
            fill="currentColor"
          />
        </svg>
         Music For Coding
      </h1>
    </a>
    <button>
      Library 
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-music "
        data-icon="music"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M511.1 367.1c0 44.18-42.98 80-95.1 80s-95.1-35.82-95.1-79.1c0-44.18 42.98-79.1 95.1-79.1c11.28 0 21.95 1.92 32.01 4.898V148.1L192 224l-.0023 208.1C191.1 476.2 149 512 95.1 512S0 476.2 0 432c0-44.18 42.98-79.1 95.1-79.1c11.28 0 21.95 1.92 32 4.898V126.5c0-12.97 10.06-26.63 22.41-30.52l319.1-94.49C472.1 .6615 477.3 0 480 0c17.66 0 31.97 14.34 32 31.99L511.1 367.1z"
          fill="currentColor"
        />
      </svg>
    </button>
  </nav>
</div>
`;
